components/brave_wallet/browser/internal/hd_key.cc

```diff
@@ -430,6 +431,35 @@ std::vector<uint8_t> HDKey::GetPublicKeyFromX25519_XSalsa20_Poly1305() const {
   return public_key;
 }
 
+std::vector<uint8_t> HDKey::DecryptCipherFromX25519_XSalsa20_Poly1305(
+    const std::string& version,
+    const std::vector<uint8_t>& nonce,
+    const std::vector<uint8_t>& ephemeral_public_key,
+    const std::vector<uint8_t>& ciphertext) const {
+  // Only x25519-xsalsa20-poly1305 is supported by MM at the time of writing
+  if (version != "x25519-xsalsa20-poly1305")
+    return std::vector<uint8_t>();
+  if (nonce.size() != crypto_box_curve25519xsalsa20poly1305_tweet_NONCEBYTES)
+    return std::vector<uint8_t>();
+  if (ephemeral_public_key.size() !=
+      crypto_box_curve25519xsalsa20poly1305_tweet_PUBLICKEYBYTES)
+    return std::vector<uint8_t>();
+  if (private_key_.size() !=
+      crypto_box_curve25519xsalsa20poly1305_tweet_SECRETKEYBYTES)
+    return std::vector<uint8_t>();
+
+  std::vector<uint8_t> padded_ciphertext = ciphertext;
+  padded_ciphertext.insert(padded_ciphertext.begin(), 16, 0);
+  std::vector<uint8_t> padded_plaintext(ciphertext.size() + 32);
+  if (crypto_box_open(padded_plaintext.data(), padded_ciphertext.data(),
+                      padded_ciphertext.size(), nonce.data(),
+                      ephemeral_public_key.data(), private_key_.data()) != 0)
+    return std::vector<uint8_t>();
+  std::vector<uint8_t> plaintext(padded_plaintext.cbegin() + 32,
+                                 padded_plaintext.cend() - 16);
```

Comment:
with https://github.com/brave/brave-core/pull/12840/commits/c8fd21530667c127544ea791c4b2f5615a876e40#r840908466
this should be just `padded_plaintext.cend()`

Reply:
There are always 16 0 bytes after the message and 32 before the message so I wasn't able to do that, but I did update to use the constants in 6e61215fb5780da624ae7803ae65d9b0fb6c03e7

Reply:
as per our discussion on Slack, I updated with the latest (ciphertext is always larger even for 0 byte message). Latest is in 6776bb4d16b20d0a806867457cbcadb86d1d7bd5